instrgen/docs/design.md

```diff
@@ -0,0 +1,186 @@
+# OpenTelemetry Go Source Automatic Instrumentation - design
+
+This document describes design of `instrgen` - compile time automatic instrumentation for GO programming language.
+
+## Background
+
+Usually techniques used for instrumenting programs in specific languages are connected to runtimes and type systems
+used by these languages and depend on capabilities offered by them. Methods differs depending on language properties.
+Is executed by virtual machine? Is statically or dynamically typed?
+What's matter is the injection point.
+In dynamic languages it's very easy to replace original implementation by providing a new one
+usually build on top of original one, a technique called monkey patching (look at python decorators).
+
+GO is statically typed, natively compiled programming language which means that works on bare methal without
+virtual machine in between. The only component that lives between your program and machine is operating system.
+Therefore, to instrument GO programs, injection needs to be done at different level. There are two ways
+utilized for now by opentelemetry community for GO:
+
+1) an ebpf (https://ebpf.io/) a technology that allowes "programming" kernel 
+   and can be used by native programming languages in general on platforms that provide it.
+   
+2) compile time instrumentation - used by `instrgen`
+
+## Rationale
+
+Compile time instrumentation is not different by other mechanisms described above.
+The fundamental difference is the injection point, instrumentation has to be done 
+at compile time which has cons and pros.
+
+GO is perfect language for compile time instrumentation for two main reasons:
+- everything what's needed is in standard library
+- it's easy to control compilation process due to go compiler architecture
+
+## Design
+
+As mentioned above `instrgen` utilizes toolexec compiler switch (https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4058/).
+
+**NOTE** `driver` executable needs to be accessible via $PATH
+
```

Comment:
Please add a "Scope" section here

```suggestion
### Scope
```

Include in this section the initial scope of the project and also what the scope does not include.

Based on our conversation at KubeCon the initial scope should include auto-instrumentation of ...

- HTTP handlers 
- gRPC handlers

The initial scope should not include ...

- All functions

A second phase of the project scope will include ...

- Selection mechanisms to instrument additional business logic.

Please be sure to include language stating that `instrgen` is meant to help start the instrumentation process. It is a utility to help take away all the tedious boilerplate work that needs to be done when starting down the manual instrumentation journey.

cc @dashpole please add any other thoughts I missed.

Reply:
Thx. I planned to add something like scope section.